### PR TITLE
Modified public API to allow unit testing of User class

### PR DESCRIPTION
**Pull request by duncanpMS**
The public API exposes the _User_ class which in turn exposes properties a developer might reasonably want to use in their app such as _Name_ and _DisplayableId_ - which in turn means they will want to be able to unit test their app.

### CODE REVIEW

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 1]
```diff
@@ -112,7 +113,7 @@ public IEnumerable<User> Users
                 {
                     RequestContext requestContext = new RequestContext(CorrelationId);
                     requestContext.Logger.Info("Token cache is null or empty");
-                    return new List<User>();
+                    return Enumerable.Empty<User>();
```
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/ClientApplicationBase.cs]: Miscellaneous: use _Enumerable.Empty<>_ for lists that should not have any members.

File: src/Microsoft.Identity.Client/User.cs [hunk 2]
```diff
@@ -47,37 +46,34 @@ internal User(User other)
             IdentityProvider = other.IdentityProvider;
         }
 
+        public User(string identifier, string displayableId, string name, string identityProvider)
```
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/User.cs]: Changed the _Create_ method to be a public constructor.
I'm guessing that the _identifier_ value is required, so I've moved it first (it's usual to have required parameter appearing before optional parameters).

File: src/Microsoft.Identity.Client/User.cs [hunk 3]
```diff
@@ -47,37 +46,34 @@ internal User(User other)
...
+        {
+            if (string.IsNullOrWhiteSpace(identifier))
+            {
+                throw new ArgumentNullException(nameof(identifier));
+            }
+
+            DisplayableId = displayableId;
+            Name = name;
+            IdentityProvider = identityProvider;
```
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/User.cs]: @kpanwar are any of the other parameters required, or can they be null?

File: src/Microsoft.Identity.Client/User.cs (hunk 2 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/User.cs]: We must not allow creation of user object publicially. This should be an internal detail

File: src/Microsoft.Identity.Client/User.cs (hunk 3 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/User.cs]: We should not have a public constructor in this class. This is an internal detail of the API that should be constructed internally.

Comment (duncanpMS) [inline src/Microsoft.Identity.Client/User.cs]: Users need to be able to test that their apps can use the returned _User_ objects correctly, so either they need to be able to create _User_ objects or there needs to be an _IUser_ interface. As this is now a data-only class with only simple write-once fields it didn't seem worth creating an interface. However, I'll create one if you prefer.

Comment (kpanwar) [inline src/Microsoft.Identity.Client/User.cs]: Whole point of unit testing MSAL was to get coverage for API behavior. Developer testing MSAL would either have a failure or have a auth result come back. What additional scenarios do you think the developer need to test to be able to test negative behavior? I don't see a need for user object to have an interface or public constructors

Comment (duncanpMS) [inline src/Microsoft.Identity.Client/User.cs]: These changes are not about unit testing MSAL or getting coverage of the MSAL API - they are about allow the developer to test their own application.
e.g. the application developer might use a property like _User.DisplayableId_ in their app, so they need to be able to create a dummy user that has known properties so they can test their app does the right thing with it.
They might also make a sequence of calls to MSAL - call (1) returns a user object that is later used in call (2). If they want to test their app passes the _User_ object correctly between calls then they need to be able to write a unit test that creates a dummy _User_ object.

Comment (kpanwar) [inline src/Microsoft.Identity.Client/User.cs]: They don't need to create a dummy user object to be able to mock out a behavior of acquire token API. They can pass null for all they care into the mocked method. Adding interface to user object is just another level of indirection as it does not provide behavior.

Comment (duncanpMS) [inline src/Microsoft.Identity.Client/User.cs]: It's not just about the behaviour of acquiring a token; it's also about the data that is returned.
Put yourself if the position of a developer consuming MSAL: you call the app to acquire a token, and if the call succeeds you display the _Result.User.Name_ in the UI. How do you test that the name is being displayed correctly if you can't create a user object with valid data?

Or how do they test that their app acquires a token and then later passes the returned _Result.User_ to _IClientApplicationBase.Remove(user)_? Using a null _User_ isn't sufficient because they want to check that their application is correctly passing around the expected user object. They are not testing MSAL - they are testing that their app correctly passes around the user object that is returned by their mocked version of MSAL.

Comment (kpanwar) [inline src/Microsoft.Identity.Client/User.cs]: In that case, interface is what we would need but not public constructor

Comment (duncanpMS) [inline src/Microsoft.Identity.Client/User.cs]: Ok, no problem.

File: samples/XForms/XForms.UWP/BundleArtifacts/x86.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+MainPackage=C:\repos\aad\forks\msal.net\samples\XForms\XForms.UWP\bin\x86\Debug\XForms.UWP_1.0.0.0_x86_Debug.appx
```
Comment (duncanpMS) [inline samples/XForms/XForms.UWP/BundleArtifacts/x86.txt]: Oops, didn't mean to include this. I'll remove it.
